File: src/main/org/nlogo/nvm/Procedure.java
package org.nlogo.nvm;

import java.util.ArrayList;
import java.util.List;

import org.nlogo.api.Dump;
import org.nlogo.api.SourceOwner;
import org.nlogo.api.Let;
import org.nlogo.api.Token;

public strictfp class Procedure {

  // maybe we should just use a boolean for this - ST 2/7/11
  public enum Type {
    COMMAND, REPORTER
  }

  public final Type tyype;
  public final Token nameToken;
  public final String fileName; // used by cities include-file stuff
  public final String name;
  public final String displayName;
  public int pos;
  public int endPos;
  public List<String> args = new ArrayList<String>();
  public String usableBy = "OTPL";
  public int localsCount = 0;
  public boolean topLevel = false;
  private SourceOwner owner;
  public final Procedure parent;
  public final scala.collection.mutable.ArrayBuffer<Procedure> children =
      new scala.collection.mutable.ArrayBuffer<Procedure>();

  public boolean isLambda() {
    return parent != null;
  }

  public int size; // cache args.size() for efficiency with making Activations

  // ExpressionParser doesn't know how many parameters the lambda is going to take;
  // that's determined by LambdaVisitor. so for now this is mutable - ST 2/4/11
  public final scala.collection.mutable.ArrayBuffer<Let> lambdaFormals =
      new scala.collection.mutable.ArrayBuffer<Let>();

  public Let getLambdaFormal(int n, Token token) {
    while (lambdaFormals.size() < n) {
      lambdaFormals.$plus$eq(
          new Let("?" + n, token.startPos(), token.endPos(),
              java.util.Collections.<Let>emptyList()));
    }
    return lambdaFormals.apply(n - 1);
  }

  public final List<Let> lets = new ArrayList<Let>();
  // each Int is the position of that variable in the procedure's args list
  public final scala.collection.mutable.HashMap<Let, Integer> alteredLets =
      new scala.collection.mutable.HashMap<Let, Integer>();

  public Command[] code = new Command[0];

  public Procedure(Type tyype, Token nameToken, String name, scala.Option<String> displayName, Procedure parent) {
    this.tyype = tyype;
    this.nameToken = nameToken;
    this.name = name;
    this.fileName = nameToken.fileName();
    this.parent = parent;
    this.displayName = buildDisplayName(displayName);
  }

  private String buildDisplayName(scala.Option<String> displayName) {
    return isLambda() ?
        "(command task from: " + parent.displayName + ")" :
        (displayName.isDefined() ? displayName.get() : ("procedure " + getNameAndFile()));
  }

  private String getNameAndFile() {
    if (fileName == null || fileName.length() == 0) {
      return name;
    }
    return name + " (" + fileName + ")";
  }

  public Syntax syntax() {
    int[] right = new int[(args.size() - localsCount)];
    for (int i = 0; i < right.length; i++) {
      right[i] = Syntax.TYPE_WILDCARD;
    }
    switch (tyype) {
      case COMMAND:
        return Syntax.commandSyntax(right);
      case REPORTER:
        return Syntax.reporterSyntax(right, Syntax.TYPE_WILDCARD);
      default:
        throw new IllegalStateException();
    }
  }

  @Override
  public String toString() {
    return super.toString() +
        "[" + name + ":" + Dump.list(args) + ":" + usableBy + "]";
  }

  public String dump() {
    StringBuilder buf = new StringBuilder();
    boolean indent = isLambda();
    if (indent) {
      buf.append("   ");
    }
    buf.append(displayName);
    if (parent != null) {
      buf.append(":" + parent.displayName);
    }
    buf.append(":" + Dump.list(args) + "{" + usableBy + "}:\n");
    for (int i = 0; i < code.length; i++) {
      if (indent) {
        buf.append("   ");
      }
      Command command = code[i];
      buf.append("[" + i + "]");
      buf.append(command.dump(indent ? 6 : 3));
      buf.append("\n");
    }
    for (Procedure p : org.nlogo.util.JCL.toJavaIterable(children)) {
      buf.append("\n");
      buf.append(p.dump());
    }
    return buf.toString();
  }

  public void init(Workspace workspace) {
    size = args.size();
    for (int i = 0; i < code.length; i++) {
      code[i].init(workspace);
    }
    for (Procedure p : org.nlogo.util.JCL.toJavaIterable(children)) {
      p.init(workspace);
    }
  }

  public SourceOwner getOwner() {
    return owner;
  }

  public void setOwner(SourceOwner owner) {
    this.owner = owner;
    for (Procedure p : org.nlogo.util.JCL.toJavaIterable(children)) {
      p.setOwner(owner);
    }
  }
}

File: src/main/org/nlogo/nvm/Procedure.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm

import org.nlogo.api

class Procedure(
  val isReporter: Boolean,
  val name: String,
  val nameToken: api.Token,
  val argTokens: Seq[api.Token],
  _displayName: Option[String] = None,
  val parent: Procedure = null) {

  val filename = nameToken.filename; // used by cities include-file stuff
  val displayName = buildDisplayName(_displayName)
  var pos: Int = 0
  var end: Int = 0
  var usableBy = "OTPL"
  var localsCount = 0
  var topLevel = false
  private var _owner: api.SourceOwner = null
  val children = collection.mutable.Buffer[Procedure]()
  var args = Vector[String]()
  def isTask = parent != null

  // cache args.size() for efficiency with making Activations
  var size = 0

  // ExpressionParser doesn't know how many parameters the task is going to take;
  // that's determined by TaskVisitor. so for now this is mutable - ST 2/4/11
  val taskFormals = collection.mutable.Buffer[api.Let]()

  def getTaskFormal(n: Int, token: api.Token): api.Let = {
    while (taskFormals.size < n)
      taskFormals += api.Let("?" + n, token.start, token.end)
    taskFormals.last
  }

  var code = Array[Command]()

  private def buildDisplayName(displayName: Option[String]): String =
    if (isTask)
      "(command task from: " + parent.displayName + ")"
    else {
      def nameAndFile =
        Option(filename)
          .filter(_.nonEmpty)
          .map(name + " (" + _ + ")")
          .getOrElse(name)
      displayName.getOrElse("procedure " + nameAndFile)
    }

  def syntax: api.Syntax = {
    val right = Array.fill(args.size - localsCount)(api.Syntax.WildcardType)
    if (isReporter)
      api.Syntax.reporterSyntax(right, api.Syntax.WildcardType)
    else
      api.Syntax.commandSyntax(right)
  }

  override def toString =
    super.toString + "[" + name + ":" + args.mkString("[", " ", "]") +
      ":" + usableBy + "]"

  def dump: String = {
    val buf = new StringBuilder
    val indent = isTask
    if (indent)
      buf ++= "   "
    if (isReporter)
      buf ++= "reporter "
    buf ++= displayName
    if (parent != null)
      buf ++= ":" + parent.displayName
    buf ++= ":"
    buf ++= args.mkString("[", " ", "]")
    buf ++= "{" + usableBy + "}:\n"
    for (i <- code.indices) {
      if (indent)
        buf ++= "   "
      val command = code(i)
      buf ++= "[" + i + "]"
      buf ++= command.dump(if (indent) 6 else 3)
      buf ++= "\n"
    }
    for (p <- children) {
      buf ++= "\n"
      buf ++= p.dump
    }
    buf.toString
  }

  def init(workspace: Workspace) {
    size = args.size
    code.foreach(_.init(workspace))
    children.foreach(_.init(workspace))
  }

  def owner = _owner
  def owner_=(owner: api.SourceOwner) {
    _owner = owner
    children.foreach(_.owner = owner)
  }

}
